lib/bunch/binary.ex

```diff
@@ -5,6 +5,74 @@ defmodule Bunch.Binary do
 
   use Bunch
 
+  @doc """
+  Returns a tuple `{list, rest}` where each element of the list is the result of
+  invoking `f` on a part of binary, and rest is the unprocessed leftover.
+
+  Function `f` is always given entire non-consumed part binary and is supposed to
+  return a subsequent output element and the remaining part of the binary to be
+  processed by further calls to the function. Once the remaining data becomes
+  ineligible for processing, the function should return `:halt`.
+
+  ## Examples
+
+      iex> f = fn
+      iex>   <<size, content::binary-size(size), rest::binary>> -> {content, rest}
+      iex>   _binary -> :halt
+      iex> end
+      iex> #{inspect(__MODULE__)}.map_while(<<2, "ab", 3, "cde", 4, "fghi">>, f)
+      {~w(ab cde fghi), <<>>}
+      iex> #{inspect(__MODULE__)}.map_while(<<2, "ab", 3, "cde", 4, "fg">>, f)
+      {~w(ab cde), <<4, "fg">>}
+
+  """
+  @spec map_while(binary, (binary -> {a, binary} | :halt)) :: {[a], binary} when a: any
+  def map_while(binary, f) do
+    do_map_while(binary, f, [])
+  end
+
+  defp do_map_while(binary, f, acc) do
+    case f.(binary) do
+      {value, rest} -> do_map_while(rest, f, [value | acc])
+      :halt -> {Enum.reverse(acc), binary}
+    end
+  end
+
+  @doc """
+  The version of `map_while/2` that accepts `:ok` and `:error` return tuples.
+
+  Mapping is continued as long as `:ok` tuples are returned, upon `:error` it breaks
```

Comment:
`It breaks` suggest it's bugged :stuck_out_tongue: I'd rather use stops processing/mapping/matching/whatever